allennlp/training/tensorboard_writer.py

```diff
@@ -17,30 +17,45 @@ class TensorboardWriter(FromParams):
 
     # Parameters
 
-    get_batch_num_total : Callable[[], int]
-        A thunk that returns the number of batches so far. Most likely this will
-        be a closure around an instance variable in your `Trainer` class.
     serialization_dir : str, optional (default = None)
         If provided, this is where the Tensorboard logs will be written.
     summary_interval : int, optional (default = 100)
         Most statistics will be written out only every this many batches.
     histogram_interval : int, optional (default = None)
         If provided, activation histograms will be written out every this many batches.
         If None, activation histograms will not be written out.
+        When this parameter is specified, the following additional logging is enabled:
+            * Histograms of model parameters
+            * The ratio of parameter update norm to parameter norm
+            * Histogram of layer activations
+        We log histograms of the parameters returned by
+        `model.get_parameters_for_histogram_tensorboard_logging`.
+        The layer activations are logged for any modules in the `Model` that have
+        the attribute `should_log_activations` set to `True`.  Logging
+        histograms requires a number of GPU-CPU copies during training and is typically
```

Comment:
> requires a number of GPU-CPU copies during training

While this is true and completely fine to state, it makes me remember that using tqdm after every batch to show the training status has also the same effect.

Reply:
Can you elaborate on that? Doesn't `tqdm` just copy a tiny tensor, instead of all the parameters?

Reply:
This didn't use to be true of metrics, but now metrics can live entirely on the GPU.  If they do, then you'll be doing GPU/CPU copies of the metric values when reporting metrics during training / validation.  This is *much* lower bandwidth transfer than doing the same for weights and activations.  Not sure how much impact it has on runtimes.

Reply:
> Can you elaborate on that? Doesn't `tqdm` just copy a tiny tensor, instead of all the parameters?

Yes. When the GPU operations are executed with the current data batch (supposing GPU usage), they live in GPU memory and never do GPU-CPU sync, and are executed async actually because they're enqueued by PyTorch transparently. When any GPU variable is read and needs to be used in CPU memory, the CUDA device is gonna be synchronized (waiting for the necessary operations to finish execution) and transferred to CPU memory, to then continue with the execution. This happens when displaying the metrics with tqdm:

1. The batch is executed on GPU. The metrics are computed on GPU.
2. The value is copied to CPU to show in tqdm.
3. Next batch...

The next batch waits for the transfer in the middle. Which is different from not waiting for GPU-CPU transfers (which are expensive relative to GPU execution) between many batch executions. It could be solved with some `log_every` option or similar.

Btw, it'd be good to measure this somehow (e.g, show there are no transfers with some profiling, doing transfers on purpose to see how slow it is).

As a side note, I'm not 100% confident this is how it works, but I'm sure enough to raise the concern.

Relevant documentation: [Asynchronous execution - CUDA Semantics - PyTorch master documentation](https://pytorch.org/docs/stable/notes/cuda.html#asynchronous-execution).